VisualPinball.Unity/VisualPinball.Unity/VPT/Mech/ScoreMotorComponent.cs

```diff
@@ -0,0 +1,293 @@
+	public delegate void ScoreMotorAddPointsCallback(float points);
+
+	[AddComponentMenu("Visual Pinball/Game Item/Score Motor")]
+	public class ScoreMotorComponent : MonoBehaviour, ISwitchDeviceComponent
+	{
+		public const int MaxIncrease = 5;
+
+		[Unit("\u00B0")]
+		[Tooltip("The total number of degrees in one turn.")]
+		public int Degrees = 120;
+
+		[Unit("ms")]
+		[Tooltip("Amount of time, in milliseconds to move one turn.")]
+		public int Duration = 769;
+
+		[Tooltip("The total number of steps per turn.")]
+		[Min(MaxIncrease)]
+		public int Steps = 6;
+
+		[Tooltip("Disable to allow single point scores while score motor running.")]
+		public bool BlockScoring = true;
+
+		public List<ScoreMotorTiming> ScoreMotorTimingList = new List<ScoreMotorTiming>() {
+			new ScoreMotorTiming(),
+			new ScoreMotorTiming(),
+			new ScoreMotorTiming(),
+			new ScoreMotorTiming(),
+			new ScoreMotorTiming()
+		};
+
+		public const string MotorRunningSwitchItem = "motor_running_switch";
+		public const string MotorStepSwitchItem = "motor_step_switch";
+
+		public IEnumerable<GamelogicEngineSwitch> AvailableSwitches => new[] {
+			new GamelogicEngineSwitch(MotorRunningSwitchItem)
+			{
+				Description = "Motor Running Switch"
+			},
+			new GamelogicEngineSwitch(MotorStepSwitchItem)
+			{
+				Description = "Motor Step Switch",
+				IsPulseSwitch = true
+			}
+		};
+
+		public SwitchDefault SwitchDefault => SwitchDefault.NormallyOpen;
+		IEnumerable<GamelogicEngineSwitch> IDeviceComponent<GamelogicEngineSwitch>.AvailableDeviceItems => AvailableSwitches;
+
+		public event EventHandler<SwitchEventArgs2> OnSwitchChanged;
+
+		private static readonly Logger Logger = LogManager.GetCurrentClassLogger();
+
+		private int DegreesPerStep => Degrees / Steps;
+		private float DegreesPerSecond => Degrees / (Duration / 1000f);
+
+		private bool _running;
+
+		private float _time;
+		private int _pos;
+
+		private ScoreMotorMode _mode;
+		private string _id;
+		private int _increase;
+
+		private float _score;
+		private ScoreMotorResetCallback _resetCallback;
+
+		private float _points;
+		private ScoreMotorAddPointsCallback _addPointsCallback;
+
+		#region Runtime
+
+		private void Awake()
+		{
+			GetComponentInParent<Player>().RegisterScoreMotorComponent(this);
+		}
+
+		private void Switch(string id, bool isClosed)
+		{
+			OnSwitchChanged?.Invoke(this, new SwitchEventArgs2(id, isClosed));
+		}
+
+		private void Update()
+		{
+			if (!_running) {
+				return;
+			}
+
+			_time += Time.deltaTime;
+
+			int currentPos = (int)(DegreesPerSecond * _time);
+
+			while (_pos <= currentPos && _pos < Degrees) {
+				AdvanceMotor();
+			}
+
+			if (_pos >= Degrees) {
+				if (_mode == ScoreMotorMode.Reset) {
+					if (_score > 0) {
+						_time = 0;
+						_pos = 0;
+
+						AdvanceMotor();
+
+						return;
+					}
+				}
+
+				StopMotor();
+			}
+		}
+
+		public void Reset(string id, float score, ScoreMotorResetCallback callback)

```

Comment:
I'll rename this to `ResetScore`, since [Reset is already taken](https://docs.unity3d.com/ScriptReference/MonoBehaviour.Reset.html).